Design/Instrument-2.0/Instrument_Ray_Tracer_2.md

```diff
@@ -0,0 +1,68 @@
+
+ * `void trace(const Kernel::V3D &dir) const`
+ * `void traceFromSample(const Kernel::V3D &dir) const`
+ * `Links getResults() const`
+ * `IDetector_const_sptr getDetectorResult() const`
+ * `void fireRay(Track &testRay) const`
+
+#### Possible New Methods
+Although the headers do not explicity have anything to do with `Instrument`, many of the method bodies make use of the member variable `m_instrument` to carry out various tasks. This variable is created and initialised in the constructor. From looking at the implementation in [`InstrumentRayTracer.cpp`](https://github.com/BhuvanBezawada/mantid/blob/master/Framework/Geometry/src/Objects/InstrumentRayTracer.cpp), it seems that the constructor is the one place where there is a dependence on `Instrument` and this is probably the one area that needs to be fully rewritten. 
+
+## The Constructor
+To rewrite the constructor it may be enough to pass in a `const` reference to a `ComponentInfo` object. This would ensure that the `InstrumentRayTracer` works as intended but there is a chance that undefined behaviour could arise if scoping is used.
+
+```c++
+
+// This example would work as intended
+InstrumentRayTracer2* ptr;
+auto ws = CreateWorkspace();
+ptr = new InstrumentRayTracer2(ws.componentInfo);
+ptr.trace(V3D(0,0,0));
+
+// With the following scoping, there would be undefined behaviour
+InstrumentRayTracer2* ptr;
+{ 
+  auto ws = CreateWorkspace();
+  ptr = new InstrumentRayTracer2(ws.componentInfo);
+}
+ptr.trace(V3D(0,0,0));
+
+```
+
+#### `ComponentInfo` Copy Implementation 
+One other possible implementation could be that a copy of a `ComponentInfo` object is passed to the constructor. ComponentInfo is also cheap to copy because in the [copy constructor](https://github.com/mantidproject/mantid/blob/8ec802f56c5db2261a0f9502f30f67fe42530d62/Framework/Geometry/src/Instrument/ComponentInfo.cpp#L88) shared pointers are created to the orginal copy's data. This should also ensure that the undefined behaviour mentioned above should not happen.
+
+#### Classless Implementation
+Another idea could be to make the methods of `InstrumentRayTracer 2.0` a set of free functions. Each of the functions would need to take in a `ComponentInfo` and any other variables required to carry out the correct procedure. This approach would definitely eliminate the passing around of the `InstrumentRayTracer` object that currently happens. Also, there are not many variables that would need to be passed to each of the methods meaning calls to the functions would remain largely the same.
+
+## Usage of `ComponentInfo`
+If all goes well with the plan to use `ComponentInfo` where possible, it is very likely that the code to use `InstrumentRayTracer 2.0` will be very similar to the current method of using it. Possible methods from `ComponentInfo` that could be used in `InstrumentRayTracer 2.0` might include:
+
```

Comment:
Is this an exhaustive list? This is one thing we need to be definite about. i.e. is it possible to use `ComponentInfo` for the job of `Instrument` without changes to current `ComponentInfo`?

Reply:
I'll make the list more comprehensive so we have a clearer idea of what we can use.

Reply:
Unless we are using MDLeanEvents, all the MDEvents have their detectorID.  Instead of ray tracing the box center back to the detector, why don't we use the detectorID for some event near the center or use statistics on all the detectorIDs of events in the box?

Reply:
I think that would be an additional feature/change that is outside the scope of what Bhuvan is considering here. Bhuvan is attempting to factor out the `Instrument` object from `Peak` in this first pass. `ComponentInfo` still has the ability to deal with detectorID, so can do exactly what `Instrument` currently does, so the changes would still allow you to develop what you suggest Vickie.

Reply:
Note for MDEvents you also have an experiment info per event. You might have events corresponding to different experiment infos. It is possible that the instrument geometry changes as a function of some logs, so that needs to be taken into account